src/controllers/marketingEmail.controller.ts

```diff
@@ -1,61 +1,79 @@
+import { sendEmail } from "../utils/email.util";
 
 /**
- * This controller creates a marketing email
+ * This controller will send marketing emails
  *
- * @param req - Request<{}, {}, CreateMarketingEmailSchema["body"]>
- * @param {Response} res - Response - this is the response object that will be sent back to the client
+ * @param req express request
+ * @param res express response
  *
- * @author tharun634
+ * @author tharun634, aayushchugh
  */
 export const createMarketingEmailHandler = async (
 	req: Request<{}, {}, CreateMarketingEmailSchema["body"]>,
 	res: Response,
 ) => {
 	try {
-		const { subject, users } = req.body;
-		await createMarketingEmailService({
-			subject,
-			users: users.map((user) => new Types.ObjectId(user)),
-		});
-		res.status(StatusCodes.CREATED).json({
-			message: "Successfully created marketing email",
-		});
-	} catch (err) {
-		return res.status(StatusCodes.INTERNAL_SERVER_ERROR).json({
-			error: "Internal server error",
-		});
-	}
-};
+		const { subject, specificUsers, allUsers, html } = req.body;
 
-/**
- * This controller sends a marketing email
- *
- * @param req - Request<{}, {}, SendMarketingEmailSchema["body"]>
- * @param {Response} res - Response - this is the response object that will be sent back to the client
- *
- * @author tharun634
- */
-export const sendMarketingEmailHandler = async (
-	req: Request<{}, {}, SendMarketingEmailSchema["body"]>,
-	res: Response,
-) => {
-	try {
-		const { html, subject } = req.body;
-		await sendMarketingEmailService(html, subject);
-		res.status(StatusCodes.OK).json({
-			message: "Successfully sent marketing email(s)",
-		});
+		// validate request
+		if (!specificUsers && !allUsers) {
+			return res.status(StatusCodes.BAD_REQUEST).json({
+				error: "You must select either specificUsers or allUsers",
+			});
+		}
+
+		if (specificUsers && specificUsers.length > 0 && allUsers) {
+			return res.status(StatusCodes.BAD_REQUEST).json({
+				error: "You can only select one, specificUsers or allUsers",
+			});
+		}
+
+		/**
+		 * Send email to all the users those have selected the checkbox
+		 */
+		if (allUsers) {
+			const users = await findUsersService({ receiveMarketingEmails: true });
+
+			if (!users) {
+				return res.status(StatusCodes.NOT_FOUND).json({
+					error: "No user has opted to receive marketing emails",
+				});
+			}
+
+			users.forEach(async (user) => await sendEmail(user.email, subject, html));
+
+			await createMarketingEmailService({ subject, users: users.map((user) => user._id) });
+
+			return res.status(StatusCodes.OK).json({
+				message: "Email sent successfully",
+			});
+		}
+
+		/**
+		 * Send email to specific users
+		 */
+		if (specificUsers && specificUsers.length > 0) {
+			// get specific users from DB
+			const users = await Promise.all(
+				specificUsers.map((userId) => findUserByIdService(userId)),
+			);
+
+			users.forEach(async (user) => user && (await sendEmail(user.email, subject, html)));
```

Comment:
Do we have to see if the `user` has `receiveMarketingEmails: true`? Or is it assured that `specificUsers` will only be a subset of `user`'s who has `receiveMarketingEmails: true`?

Reply:
Fuck, i forgot to add validation for that